File: src/lib/plan/rules.ts
/* ==========================================================================
   Plan rules: the published formulas the model was trained on.

   A line-for-line port of ml/rules.py, tested against it (parity.json). The
   app needs these even with a model in front of them, for three reasons:

     - BMR and TDEE feed the safety clamps, which are code, never the model
     - anyone outside the range the model was trained on gets these instead,
       because out of range the model is off by ~300 kcal and cannot say so
     - the training split's rep ranges and cardio are rules, not predictions
   ========================================================================== */

export type Sex = "male" | "female";
export type Goal = "cut" | "maintain" | "bulk";
export type Experience = "beginner" | "intermediate" | "advanced";
export type Equipment = "none" | "dumbbell" | "gym" | "full";

/** Conditions that change the numbers. Fed to the model. */
export const MODIFY_CONDITIONS = ["hypothyroid", "pcos", "insulin_resistance", "hypertension"] as const;
/** Conditions that must not receive a generated plan at all. */
export const REFUSE_CONDITIONS = [
  "pregnancy",
  "breastfeeding",
  "ckd",
  "type1_diabetes",
  "eating_disorder_history",
  "cancer_treatment",
] as const;
export const CONDITIONS = [...MODIFY_CONDITIONS, ...REFUSE_CONDITIONS] as const;
export type Condition = (typeof CONDITIONS)[number];

export const INJURIES = ["knee", "shoulder", "lower_back", "elbow"] as const;
export type Injury = (typeof INJURIES)[number];

export const KCAL_PER_KG_FAT = 7700;
export const ACTIVITY: Record<number, number> = { 0: 1.2, 1: 1.2, 2: 1.375, 3: 1.375, 4: 1.55, 5: 1.55, 6: 1.725, 7: 1.725 };
const CUT_BANDS: Record<Sex, [number, number][]> = {
  male: [[30, 0.25], [20, 0.2], [12, 0.15], [0, 0.1]],
  female: [[38, 0.25], [28, 0.2], [20, 0.15], [0, 0.1]],
};
const BULK_BANDS: Record<Sex, [number, number][]> = {
  male: [[20, 0.05], [15, 0.1], [0, 0.15]],
  female: [[30, 0.05], [25, 0.1], [0, 0.15]],
};
export const KCAL_FLOOR: Record<Sex, number> = { male: 1500, female: 1200 };
export const MAX_WEEKLY_LOSS_FRACTION = 0.01;
export const MAX_WEEKLY_GAIN_FRACTION = 0.005;
const PROTEIN_G_PER_KG_LEAN: Record<Goal, number> = { cut: 2.2, maintain: 2.0, bulk: 1.8 };
const FAT_G_PER_KG_BODY_MIN = 0.8;

const TDEE_SCALE: Partial<Record<Condition, number>> = { hypothyroid: 0.9 };
const CARB_SHIFT: Partial<Record<Condition, number>> = { pcos: -0.15, insulin_resistance: -0.1 };
/** Not a calorie change: a ceiling the food selector must respect. */
export const SODIUM_MG_MAX: Partial<Record<Condition, number>> = { hypertension: 1500 };

export const VOLUME: Record<Experience, Record<string, number>> = {
  beginner: { chest: 8, back: 10, quads: 8, hams: 6, glutes: 8, delts: 8, arms: 6, calves: 6 },
  intermediate: { chest: 12, back: 14, quads: 12, hams: 9, glutes: 10, delts: 12, arms: 10, calves: 8 },
  advanced: { chest: 16, back: 18, quads: 16, hams: 12, glutes: 12, delts: 16, arms: 14, calves: 10 },
};
export const MUSCLES = Object.keys(VOLUME.beginner);
export const REPS: Record<Goal, [number, number]> = { cut: [8, 12], maintain: [6, 12], bulk: [6, 10] };

export const SPLIT_LABEL: Record<string, string> = {
  full_body_x2: "Full body, twice a week",
  full_body_x3: "Full body, three times a week",
  upper_lower_x2: "Upper and lower, twice each",
  ppl_upper_lower: "Push, pull, legs, upper, lower",
  ppl_x2: "Push, pull, legs, twice through",
};

/* ---------------------------------------------------------------- body */

export const katchMcArdleBmr = (leanKg: number) => 370 + 21.6 * leanKg;

export function mifflinBmr(weightKg: number, heightCm: number, age: number, sex: Sex): number {
  const base = 10 * weightKg + 6.25 * heightCm - 5 * age;
  return sex === "male" ? base + 5 : base - 161;
}

/** Rough body fat from BMI, used only to pick a deficit band. */
export const deurenbergBodyFat = (bmi: number, age: number, sex: Sex) =>
  1.2 * bmi + 0.23 * age - 10.8 * (sex === "male" ? 1 : 0) - 5.4;

function band(value: number, bands: [number, number][]): number {
  for (const [threshold, frac] of bands) if (value >= threshold) return frac;
  return bands[bands.length - 1][1];
}

/** Round the way Python's round() does for these magnitudes. */
const r = (v: number, d: number) => Math.round(v * 10 ** d) / 10 ** d;

/* ---------------------------------------------------------------- targets */

export type TargetInput = {
  weightKg: number;
  heightCm: number;
  age: number;
  sex: Sex;
  bodyFatPct: number | null;
  days: number;
  goal: Goal;
  conditions?: readonly string[];
};

export type Targets = {
  bmr: number;
  tdee: number;
  kcal: number;
  proteinG: number;
  carbsG: number;
  fatG: number;
  deficit: number;
  weeklyKgChange: number;
  clamped: string[];
  leanKg: number;
  bodyFatPct: number;
};

/** BMR, lean mass and the body fat used for banding, for either path. */
export function body(input: TargetInput): { bmr: number; leanKg: number; bodyFatPct: number } {
  const { weightKg, heightCm, age, sex, bodyFatPct } = input;
  if (bodyFatPct !== null) {
    const leanKg = weightKg * (1 - bodyFatPct / 100);
    return { bmr: katchMcArdleBmr(leanKg), leanKg, bodyFatPct };
  }
  const bmi = weightKg / (heightCm / 100) ** 2;
  const bf = Math.max(3, Math.min(65, deurenbergBodyFat(bmi, age, sex)));
  return { bmr: mifflinBmr(weightKg, heightCm, age, sex), leanKg: weightKg * (1 - bf / 100), bodyFatPct: bf };
}

export function tdeeFor(bmr: number, days: number, conditions: readonly string[] = []): number {
  let tdee = bmr * ACTIVITY[Math.max(0, Math.min(7, days))];
  for (const c of conditions) tdee *= TDEE_SCALE[c as Condition] ?? 1;
  return tdee;
}

/** Profile in, daily numbers out, entirely by formula. Mirrors rules.targets. */
export function formulaTargets(input: TargetInput): Targets {
  const { weightKg, sex, days, goal } = input;
  const conditions = input.conditions ?? [];
  const { bmr, leanKg, bodyFatPct } = body(input);
  const tdee = tdeeFor(bmr, days, conditions);

  let kcal =
    goal === "cut"
      ? tdee * (1 - band(bodyFatPct, CUT_BANDS[sex]))
      : goal === "bulk"
        ? tdee * (1 + band(bodyFatPct, BULK_BANDS[sex]))
        : tdee;

  const clamped: string[] = [];
  if (goal === "cut") {
    if (kcal < bmr) {
      kcal = bmr;
      clamped.push("bmr");
    }
    const floor = Math.min(KCAL_FLOOR[sex], tdee);
    if (kcal < floor) {
      kcal = floor;
      clamped.push("floor");
    }
    if (kcal > tdee) {
      kcal = tdee;
      clamped.push("maintenance");
    }
  }

  let weekly = ((tdee - kcal) * 7) / KCAL_PER_KG_FAT;
  if (weekly > weightKg * MAX_WEEKLY_LOSS_FRACTION) {
    kcal = tdee - (weightKg * MAX_WEEKLY_LOSS_FRACTION * KCAL_PER_KG_FAT) / 7;
    clamped.push("loss_rate");
  } else if (-weekly > weightKg * MAX_WEEKLY_GAIN_FRACTION) {
    kcal = tdee + (weightKg * MAX_WEEKLY_GAIN_FRACTION * KCAL_PER_KG_FAT) / 7;
    clamped.push("gain_rate");
  }
  const deficit = tdee - kcal;
  weekly = (deficit * 7) / KCAL_PER_KG_FAT;

  let protein = PROTEIN_G_PER_KG_LEAN[goal] * leanKg;
  let fat = FAT_G_PER_KG_BODY_MIN * weightKg;
  let carbKcal = kcal - protein * 4 - fat * 9;

  const shift = conditions.reduce((s, c) => s + (CARB_SHIFT[c as Condition] ?? 0), 0);
  if (shift && carbKcal > 0) {
    const moved = carbKcal * -shift;
    carbKcal -= moved;
    fat += moved / 9;
  }
  if (carbKcal < 0) {
    fat = Math.max(0.5 * weightKg, (kcal - protein * 4) / 9);
    carbKcal = kcal - protein * 4 - fat * 9;
    clamped.push("fat_squeeze");
  }
  if (carbKcal < 0) {
    protein = Math.max(1.2 * leanKg, (kcal - fat * 9) / 4);
    carbKcal = Math.max(0, kcal - protein * 4 - fat * 9);
    clamped.push("protein_squeeze");
  }

  return {
    bmr: r(bmr, 1),
    tdee: r(tdee, 1),
    kcal: r(kcal, 1),
    proteinG: r(protein, 1),
    carbsG: r(carbKcal / 4, 1),
    fatG: r(fat, 1),
    deficit: r(deficit, 1),
    weeklyKgChange: r(weekly, 4),
    clamped,
    leanKg,
    bodyFatPct,
  };
}

/* ---------------------------------------------------------------- training */

export type SessionInput = {
  days: number;
  experience: Experience;
  equipment: Equipment;
  goal: Goal;
  injuries?: readonly string[];
};

const SPLITS: Record<number, string> = {
  2: "full_body_x2",
  3: "full_body_x3",
  4: "upper_lower_x2",
  5: "ppl_upper_lower",
  6: "ppl_x2",
};
const GOAL_VOLUME_SCALE: Record<Goal, number> = { cut: 0.85, maintain: 1, bulk: 1.1 };

/** Formula version of the training prescription. Mirrors rules.session. */
export function formulaSession(input: SessionInput): { split: string; weeklySets: Record<string, number> } {
  const days = Math.max(2, Math.min(6, input.days));
  const injuries = new Set(input.injuries ?? []);
  const equipScale = input.equipment === "none" ? 0.75 : input.equipment === "dumbbell" ? 0.9 : 1;
  const weeklySets: Record<string, number> = {};
  for (const [m, base] of Object.entries(VOLUME[input.experience])) {
    let v = base * GOAL_VOLUME_SCALE[input.goal] * equipScale;
    if (
      (injuries.has("knee") && ["quads", "hams", "glutes"].includes(m)) ||
      (injuries.has("shoulder") && ["chest", "delts"].includes(m)) ||
      (injuries.has("lower_back") && ["back", "hams", "glutes"].includes(m)) ||
      (injuries.has("elbow") && m === "arms")
    ) {
      v *= 0.5;
    }
    weeklySets[m] = pyRound(v);
  }
  return { split: SPLITS[days], weeklySets };
}

/** Python's round(): halves go to the even neighbour. */
export function pyRound(v: number): number {
  const f = Math.floor(v);
  const diff = v - f;
  if (Math.abs(diff - 0.5) < 1e-9) return f % 2 === 0 ? f : f + 1;
  return Math.round(v);
}

File: ml/rules.py
"""
Winter Arc plan rules.

Every number a plan needs, computed from a profile. This is the teacher: the
dataset generator calls it to produce labels, and the trained model learns to
approximate it. Ported to TypeScript later if the deterministic generator ships.

Sources for the constants are noted inline. Nothing here is fitted; it is all
published formulas and volume landmarks.
"""
from __future__ import annotations

from dataclasses import dataclass, asdict
from typing import Literal, Optional

Sex = Literal["male", "female"]
Goal = Literal["cut", "maintain", "bulk"]
Experience = Literal["beginner", "intermediate", "advanced"]
Equipment = Literal["none", "dumbbell", "gym", "full"]

# ---------------------------------------------------------------- constants

KCAL_PER_KG_FAT = 7700.0

# Activity multiplier by training days per week.
ACTIVITY = {0: 1.20, 1: 1.20, 2: 1.375, 3: 1.375, 4: 1.55, 5: 1.55, 6: 1.725, 7: 1.725}

# Deficit as a fraction of TDEE, chosen by body fat. Fatter bodies tolerate a
# steeper deficit; lean ones lose muscle if pushed.
CUT_BANDS_MALE = [(30.0, 0.25), (20.0, 0.20), (12.0, 0.15), (0.0, 0.10)]
CUT_BANDS_FEMALE = [(38.0, 0.25), (28.0, 0.20), (20.0, 0.15), (0.0, 0.10)]

# Surplus for a bulk: leaner bodies can afford a bigger one.
BULK_BANDS_MALE = [(20.0, 0.05), (15.0, 0.10), (0.0, 0.15)]
BULK_BANDS_FEMALE = [(30.0, 0.05), (25.0, 0.10), (0.0, 0.15)]

KCAL_FLOOR = {"male": 1500.0, "female": 1200.0}
MAX_WEEKLY_LOSS_FRACTION = 0.010   # 1.0% of bodyweight per week
MAX_WEEKLY_GAIN_FRACTION = 0.005   # 0.5% per week

PROTEIN_G_PER_KG_LEAN = {"cut": 2.2, "maintain": 2.0, "bulk": 1.8}
FAT_G_PER_KG_BODY_MIN = 0.8

# Weekly working sets per muscle group. Beginners start at minimum effective
# volume, advanced lifters sit near the top of the adaptive range.
VOLUME = {
    "beginner":     {"chest": 8,  "back": 10, "quads": 8,  "hams": 6,  "glutes": 8,  "delts": 8,  "arms": 6,  "calves": 6},
    "intermediate": {"chest": 12, "back": 14, "quads": 12, "hams": 9,  "glutes": 10, "delts": 12, "arms": 10, "calves": 8},
    "advanced":     {"chest": 16, "back": 18, "quads": 16, "hams": 12, "glutes": 12, "delts": 16, "arms": 14, "calves": 10},
}
MUSCLES = list(VOLUME["beginner"].keys())

# Cutting caps recoverable volume; a surplus lifts it.
VOLUME_GOAL_SCALE = {"cut": 0.85, "maintain": 1.0, "bulk": 1.10}

SPLITS = {2: "full_body_x2", 3: "full_body_x3", 4: "upper_lower_x2", 5: "ppl_upper_lower", 6: "ppl_x2"}

REPS = {"cut": (8, 12), "maintain": (6, 12), "bulk": (6, 10)}

# Conditions that change the numbers rather than only the food library.
CONDITION_MODIFIERS = {
    # Measured metabolic rate runs below prediction when thyroid output is low.
    "hypothyroid":        {"tdee_scale": 0.90},
    # Insulin resistance: same calories, carbs shifted into fat.
    "pcos":               {"carb_fraction_shift": -0.15},
    "insulin_resistance": {"carb_fraction_shift": -0.10},
    # No calorie effect, a sodium ceiling the food selector must respect.
    "hypertension":       {"sodium_mg_max": 1500},
}

# Conditions that must not receive a generated plan at all.
REFUSE = {
    "pregnancy", "breastfeeding", "ckd", "type1_diabetes",
    "eating_disorder_history", "under_18", "cancer_treatment",
}


def refuse_reason(conditions: set[str], age: int) -> Optional[str]:
    """Why this profile gets no plan, or None if it may have one."""
    if age < 18:
        return "under_18"
    for c in sorted(conditions):
        if c in REFUSE:
            return c
    return None


# ---------------------------------------------------------------- body

def katch_mcardle_bmr(lean_kg: float) -> float:
    """Preferred when body fat is known. Height and age do not enter."""
    return 370.0 + 21.6 * lean_kg


def mifflin_bmr(weight_kg: float, height_cm: float, age: int, sex: Sex) -> float:
    """Fallback when body fat is unknown."""
    base = 10.0 * weight_kg + 6.25 * height_cm - 5.0 * age
    return base + 5.0 if sex == "male" else base - 161.0


def deurenberg_bodyfat(bmi: float, age: int, sex: Sex) -> float:
    """Rough body fat from BMI, used only to pick a deficit band."""
    return 1.20 * bmi + 0.23 * age - 10.8 * (1 if sex == "male" else 0) - 5.4


def band(value: float, bands: list[tuple[float, float]]) -> float:
    for threshold, frac in bands:
        if value >= threshold:
            return frac
    return bands[-1][1]


# ---------------------------------------------------------------- targets

@dataclass
class Targets:
    bmr: float
    tdee: float
    kcal: float
    protein_g: float
    carbs_g: float
    fat_g: float
    deficit: float                 # positive = eating below maintenance
    weekly_kg_change: float
    clamped: str                   # which clamp bit, "" if none


def targets(
    *,
    weight_kg: float,
    height_cm: float,
    age: int,
    sex: Sex,
    bodyfat_pct: Optional[float],
    days: int,
    goal: Goal,
    conditions: frozenset[str] = frozenset(),
) -> Targets:
    """Profile in, daily numbers out. Every clamp is recorded, not silent."""
    if bodyfat_pct is not None:
        lean = weight_kg * (1.0 - bodyfat_pct / 100.0)
        bmr = katch_mcardle_bmr(lean)
        bf = bodyfat_pct
    else:
        bmi = weight_kg / (height_cm / 100.0) ** 2
        bf = max(3.0, min(65.0, deurenberg_bodyfat(bmi, age, sex)))
        lean = weight_kg * (1.0 - bf / 100.0)
        bmr = mifflin_bmr(weight_kg, height_cm, age, sex)

    tdee = bmr * ACTIVITY[days]
    for c in conditions:
        tdee *= CONDITION_MODIFIERS.get(c, {}).get("tdee_scale", 1.0)

    if goal == "cut":
        kcal = tdee * (1.0 - band(bf, CUT_BANDS_MALE if sex == "male" else CUT_BANDS_FEMALE))
    elif goal == "bulk":
        kcal = tdee * (1.0 + band(bf, BULK_BANDS_MALE if sex == "male" else BULK_BANDS_FEMALE))
    else:
        kcal = tdee

    clamped = []

    # Never below resting metabolic rate, never below the absolute floor, and
    # never above maintenance. That last one settles a real conflict: a small
    # body can have a TDEE under the floor, where "raise to the floor" would
    # turn a cut into a surplus that the gain cap then drags back under the
    # floor. A cut eats at most maintenance, so the floor stops at TDEE.
    if goal == "cut":
        if kcal < bmr:
            kcal, _ = bmr, clamped.append("bmr")
        floor = min(KCAL_FLOOR[sex], tdee)
        if kcal < floor:
            kcal, _ = floor, clamped.append("floor")
        if kcal > tdee:
            kcal, _ = tdee, clamped.append("maintenance")

    # Rate cap, both directions.
    delta = tdee - kcal
    weekly_kg = delta * 7.0 / KCAL_PER_KG_FAT
    max_loss = weight_kg * MAX_WEEKLY_LOSS_FRACTION
    max_gain = weight_kg * MAX_WEEKLY_GAIN_FRACTION
    if weekly_kg > max_loss:
        kcal = tdee - max_loss * KCAL_PER_KG_FAT / 7.0
        clamped.append("loss_rate")
    elif -weekly_kg > max_gain:
        kcal = tdee + max_gain * KCAL_PER_KG_FAT / 7.0
        clamped.append("gain_rate")

    delta = tdee - kcal
    weekly_kg = delta * 7.0 / KCAL_PER_KG_FAT

    # Macros, in priority order: protein, then a fat floor, carbs take the rest.
    protein = PROTEIN_G_PER_KG_LEAN[goal] * lean
    fat = FAT_G_PER_KG_BODY_MIN * weight_kg
    carb_kcal = kcal - protein * 4.0 - fat * 9.0

    shift = sum(CONDITION_MODIFIERS.get(c, {}).get("carb_fraction_shift", 0.0) for c in conditions)
    if shift and carb_kcal > 0:
        moved = carb_kcal * -shift
        carb_kcal -= moved
        fat += moved / 9.0

    # If protein and the fat floor overrun the budget, fat gives way first.
    if carb_kcal < 0:
        fat = max(0.5 * weight_kg, (kcal - protein * 4.0) / 9.0)
        carb_kcal = kcal - protein * 4.0 - fat * 9.0
        clamped.append("fat_squeeze")
    if carb_kcal < 0:
        protein = max(1.2 * lean, (kcal - fat * 9.0) / 4.0)
        carb_kcal = max(0.0, kcal - protein * 4.0 - fat * 9.0)
        clamped.append("protein_squeeze")

    return Targets(
        bmr=round(bmr, 1),
        tdee=round(tdee, 1),
        kcal=round(kcal, 1),
        protein_g=round(protein, 1),
        carbs_g=round(carb_kcal / 4.0, 1),
        fat_g=round(fat, 1),
        deficit=round(delta, 1),
        weekly_kg_change=round(weekly_kg, 4),
        clamped="|".join(clamped),
    )


# ---------------------------------------------------------------- training

@dataclass
class Session:
    split: str
    weekly_sets: dict          # muscle -> sets
    sets_per_session: float
    rep_low: int
    rep_high: int
    cardio_sessions: int


def session(*, days: int, experience: Experience, goal: Goal, equipment: Equipment,
            injuries: frozenset[str] = frozenset()) -> Session:
    """Split, weekly volume and rep range. Load is never prescribed."""
    split = SPLITS[days]
    scale = VOLUME_GOAL_SCALE[goal]

    # Bodyweight-only training cannot load the posterior chain the same way.
    equip_scale = 0.75 if equipment == "none" else 0.90 if equipment == "dumbbell" else 1.0

    weekly = {}
    for m, base in VOLUME[experience].items():
        v = base * scale * equip_scale
        # An injured region drops to maintenance volume rather than vanishing.
        if ("knee" in injuries and m in ("quads", "hams", "glutes")) or \
           ("shoulder" in injuries and m in ("chest", "delts")) or \
           ("lower_back" in injuries and m in ("back", "hams", "glutes")) or \
           ("elbow" in injuries and m in ("arms",)):
            v *= 0.5
        weekly[m] = int(round(v))

    total = sum(weekly.values())
    low, high = REPS[goal]
    # Cardio rides with the training days, never on rest days.
    cardio = days

    return Session(
        split=split,
        weekly_sets=weekly,
        sets_per_session=round(total / days, 2),
        rep_low=low,
        rep_high=high,
        cardio_sessions=cardio,
    )


def as_row(t: Targets, s: Session) -> dict:
    row = asdict(t)
    row.update({
        "split": s.split,
        "sets_per_session": s.sets_per_session,
        "rep_low": s.rep_low,
        "rep_high": s.rep_high,
        "cardio_sessions": s.cardio_sessions,
    })
    for m, v in s.weekly_sets.items():
        row[f"sets_{m}"] = v
    return row
